tesseract-server/src/handlers/diagnosis.rs

```diff
@@ -0,0 +1,174 @@
+};
+use failure::{Error, format_err};
+use futures::future::Future;
+use lazy_static::lazy_static;
+use log::*;
+use serde_qs as qs;
+use serde_derive::Deserialize;
+use url::Url;
+
+use tesseract_core::names::{Property, LevelName};
+use tesseract_core::format::{format_records, FormatType};
+use tesseract_core::{DataFrame, Column, ColumnData};
+use tesseract_core::schema::{Cube, DimensionType, Level};
+use crate::app::AppState;
+use crate::logic_layer::{LogicLayerConfig, CubeCache};
+use crate::handlers::util::{verify_api_key, format_to_content_type};
+use crate::handlers::logic_layer::{query_geoservice, GeoserviceQuery};
+
+
+/// Handles default aggregation when a format is not specified.
+/// Default format is jsonrecords.
+pub fn diagnosis_default_handler(
+    (req, _cube): (HttpRequest<AppState>, Path<()>)
+) -> ActixResult<HttpResponse>
+{
+    perform_diagnosis(req, "jsonrecords".to_owned())
+}
+
+
+/// Handles aggregation when a format is specified.
+pub fn diagnosis_handler(
+    (req, cube_format): (HttpRequest<AppState>, Path<(String)>)
+) -> ActixResult<HttpResponse>
+{
+    perform_diagnosis(req, cube_format.to_owned())
+}
+
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct DiagnosisQueryOpt {
+    pub cube: String,
+}
+
+
+pub fn perform_diagnosis(
+    req: HttpRequest<AppState>,
+    format: String,
+) -> ActixResult<HttpResponse>
+{
+    let format = format.parse::<FormatType>();
+    let format = match format {
+        Ok(f) => f,
+        Err(err) =>return Ok(HttpResponse::NotFound().json(err.to_string())),
+    };
+
+    info!("Format: {:?}", format);
+
+    let query = req.query_string();
+    let schema = req.state().schema.read().unwrap();
+    let _debug = req.state().debug;
+
+    lazy_static! {
+        static ref QS_NON_STRICT: qs::Config = qs::Config::new(5, false);
+    }
+
+    let query_opt = match QS_NON_STRICT.deserialize_str::<DiagnosisQueryOpt>(query) {
+        Ok(q) => q,
+        Err(err) => return Ok(HttpResponse::NotFound().json(err.to_string()))
+    };
+
+    let cube = match schema.get_cube_by_name(&query_opt.cube) {
+        Ok(c) => c,
+        Err(err) => return Ok(HttpResponse::NotFound().json(err.to_string()))
+    };
+
+    match verify_api_key(&req, &cube) {
+        Ok(_) => (),
+        Err(err) => return Ok(err)
+    }
+
+    let mut error_types: Vec<String> = vec![];
+    let mut error_messages: Vec<String> = vec![];
+
+    // Check for `MissingDimensionIDs`
+    // TODO: Deal with the case where there is an inline table.
+    for dimension in &cube.dimensions {
+        for hierarchy in &dimension.hierarchies {
+            let last_level: &Level = &hierarchy.levels[hierarchy.levels.len() - 1];
+
+            if let Some(ref foreign_key) = dimension.foreign_key {
+                if let Some(ref dimension_table) = hierarchy.table {
+                    let sql_str: String = format!(
+                        "SELECT DISTINCT {} FROM {} WHERE {} NOT IN (SELECT {} FROM {})",
```

Comment:
Two things: 
1. Just to double check: nothing in the inputs here is coming from the user? It's all only from schema? Just thinking about SQL injection
2. For now I think this is OK, though future because this kind of query should work in most SQL variants AFAIK, though we might want to delegate this to the backend driver in the future

Reply:
1. Yeah, the only thing we get from the user is the cube name.
2. Very good point. I think is generic enough that it would work for most, if not all, variants we support. But as we add more checks we should definitely refactor this.